Add IT that ALTERs right before Reader fetches next page

Bouncheck opened the pull request:
Adds a few integration tests along with necessary setup.
The changes are listed below.

Adjust classifiers in Driver3WorkerCQL:

Reader and worker classes were made non-final and protected if previously
private. This is necessary in order to implement required mocks without
use of reflection.
Methods `findNext` method, `ResultSet rs`, `lastChangeId` within Driver3Reader
are now protected instead of private too.

Add MockDriver3WorkerCQL:

Adds the mock version of Driver3WorkerCQL.
This mock is created with singular purpose of using it in tests where we
alter the table while consumer is working.
The mock adds ReentrantReadWriteLocks that allow to block the readers
in specific moments. In the case of planned tests we mainly  want to block
before fetching the next page.

Add abstract `AlterTableBase`:

Adds abstract class which outlines the test pattern, where we alter the table
while the CDCConsumer is running.
It allows for overriding what should be the table, keyspace, initial DDL,
background task for data generation, what kind of alteration should be used,
the correct RawChange schema before the alteration and after,
the RawChangeConsumer used.

Add several alter table tests:

Adds `AlterAddColIT` (adds column before all Tasks fetch next page;
also has add column before reading next row variant),
adds `AlterUpdateUdtIT` (adds new field to UDT before next page),
adds `AlterDropColIT` (drops column before next page),
adds `AlterReAddColIT` (drops then adds the same column with different type).
All tests extend `AlterTableBase` and use the same template, but different setup.

Add minor fixes to AlterTableIT:

This is an old test that was left mostly as is. The minor adjustments are
using logger instead of printStackTrace and gracefully stopping the consumer.

Add maven-jar-plugin to driver3 module:
Test-jar is now created to allow scylla-cdc-lib module to use mock class from 
driver3 module.

Addresses #133.

dkropachev reviewed (changes requested):
Let's add separate tests for Alter table on new page and on new row.

Bouncheck commented:
Found one bug in the reworked test yesterday, but some other bug still remains.

dkropachev commented:
> Found one bug in the reworked test yesterday, but some other bug still remains.

Please include fixes into this PR, update it's name and description to reflect that.

Bouncheck commented:
The initial difference came from my local cache, but after fixing that other bugs surfaced. All should be fixed now.
Fixed ResultSets not being overwritten in mock, fixed concurrency issues stemming from having too many tasks queued - after unblocking all readers it was possible that some other task not yet moving to the next page would be the first to pass the message to consumer.
Test was slightly reworked, now every row is checked rather than first and first after alter. Sleep and wait times were adjusted.
Addressed review comments, made more Driver3Reader fields protected in order to drop more reflection code.

dkropachev commented on scylla-cdc-driver3/src/test/java/com/scylladb/cdc/cql/driver3/MockDriver3WorkerCQL.java:
```diff
@@ -0,0 +1,95 @@
+
+        @Override
+        protected void findNext(CompletableFuture<Optional<RawChange>> fut) {
+            CompletableFuture.runAsync( () -> {
```
I looked closer at findNext, and now I don't think that we need to run this guy in separate feature.
Only place it tends to do something async is via `Futures.addCallback`, but it uses `MoreExecutors.directExecutor()`, so it is completely sync function, therefore we can and should dump `CompletableFuture.runAsync( () -> {` here

dkropachev commented on scylla-cdc-lib/src/test/java/com/scylladb/cdc/lib/AlterTableIT.java:
```diff
@@ -167,4 +178,193 @@ private synchronized void printDetails(RawChange change) {
+          cqlField.set(builder, mockWorkerCQL);
+        } catch (Exception e) {
+          throw new RuntimeException("Failed to replace Driver3WorkerCQL with MockDriver3WorkerCQL", e);
+        }
```
Can you please find better way to do that, without hacking into instances.
If necessary add additional API to builders.

Bouncheck replied:
It will require modifying the accesses in production code, but I can.

dkropachev commented on scylla-cdc-lib/src/test/java/com/scylladb/cdc/lib/AlterTableIT.java:
```diff
@@ -167,4 +178,193 @@ private synchronized void printDetails(RawChange change) {
+                  + " KEY (column1, column2)) WITH cdc = {'enabled': 'true'};",
+              keyspace, table));
+
+      AtomicInteger counter = new AtomicInteger(0);
```
```suggestion
      AtomicInteger updatesCount = new AtomicInteger(0);
```

dkropachev commented on scylla-cdc-lib/src/test/java/com/scylladb/cdc/lib/AlterTableIT.java:
```diff
@@ -167,4 +178,193 @@ private synchronized void printDetails(RawChange change) {
+      AtomicBoolean consumedChangeAfterAlter = new AtomicBoolean(false);
+      AtomicLong lastConsumedTime = new AtomicLong(0);
+      AtomicReference<RawChange> firstChange, firstChangeAfterFetchingNewPage;
+      firstChange = new AtomicReference<>(null);
```
```suggestion
      changeBeforeAlter = new AtomicReference<>(null);
```

Bouncheck replied:
New version has `rawChangesBeforeAlter` and `rawChangesAfterAlter` lists

dkropachev commented on scylla-cdc-lib/src/test/java/com/scylladb/cdc/lib/AlterTableIT.java:
```diff
@@ -167,4 +178,193 @@ private synchronized void printDetails(RawChange change) {
+        // and consumer has not consumed any change for the last 10 seconds
+        ReentrantReadWriteLock finalNextPageLock = nextPageLock;
+        Awaitility.await().atMost(60, TimeUnit.SECONDS).until(() -> finalNextPageLock.hasQueuedThreads() && (System.currentTimeMillis() - lastConsumedTime.get() > 10000));
+        ResultSet rs = session.execute(String.format("ALTER TABLE %s.%s " + "ADD column4 int;", keyspace, table));
```
Can we also have tests on:
1. Updating UDT
2. Removing column
3. Changing column type (by removing and adding column with the same name, but different type)

It would be great if you can generalize the test

Bouncheck replied:
All tests listed here were implemented

Bouncheck commented on scylla-cdc-lib/src/test/java/com/scylladb/cdc/lib/AlterTableIT.java:
```diff
@@ -167,4 +178,193 @@ private synchronized void printDetails(RawChange change) {
+          }
+          return false;
+        });
+        nextPageLock.writeLock().lock();
```
At this point the counter surely has ticked because I check that it's greater than 15 before this line.
It's not an issue. It's used for producing messages, not consuming.

dkropachev replied:
Oh, so `Awaitility.await().atMost` will fail then, then i am wrong.

Bouncheck commented on scylla-cdc-lib/src/test/java/com/scylladb/cdc/lib/AlterTableIT.java:
```diff
@@ -167,4 +178,193 @@ private synchronized void printDetails(RawChange change) {
+      verifySchemaBeforeAlter(firstChange.get());
+      verifySchemaAfterAlter(firstChangeAfterFetchingNewPage.get());
+
+      assertFalse(failedDueToInvalidTypeEx.get());
```
this is a leftover, will remove that

Bouncheck commented on scylla-cdc-lib/src/test/java/com/scylladb/cdc/lib/AlterTableIT.java:
```diff
@@ -167,4 +178,193 @@ private synchronized void printDetails(RawChange change) {
+      AtomicLong lastConsumedTime = new AtomicLong(0);
+      AtomicReference<RawChange> firstChange, firstChangeAfterFetchingNewPage;
+      firstChange = new AtomicReference<>(null);
+      firstChangeAfterFetchingNewPage = new AtomicReference<>(null);
```
New version has `rawChangesBeforeAlter` and `rawChangesAfterAlter` lists

Bouncheck commented on scylla-cdc-lib/src/test/java/com/scylladb/cdc/lib/AlterTableIT.java:
```diff
@@ -167,4 +178,193 @@ private synchronized void printDetails(RawChange change) {
+          } catch (Exception e) {
+            return false;
+          }
+        });
```
Removed those waits

Bouncheck commented on scylla-cdc-lib/src/test/java/com/scylladb/cdc/lib/AlterTableIT.java:
```diff
@@ -167,4 +178,193 @@ private synchronized void printDetails(RawChange change) {
+              keyspace, table));
+
+      AtomicInteger counter = new AtomicInteger(0);
+      AtomicBoolean taskShouldStop = new AtomicBoolean(false);
```
new version has `datagenShouldStop` and other workload generating things are prefixed with `datagen`

dkropachev commented on scylla-cdc-lib/src/test/java/com/scylladb/cdc/lib/AlterUpdateUdtIT.java:
```diff
@@ -0,0 +1,144 @@
+  }
+
+  @Override
+  public Runnable createDatagenTask() {
```
`data generator/generation` is a bit weird, could please use something that fit better: data writer, table writer, record insert ...

Bouncheck commented on scylla-cdc-lib/src/test/java/com/scylladb/cdc/lib/AlterTableBase.java:
```diff
@@ -0,0 +1,247 @@
+    getDriverSession().execute(createTableQuery());
+  }
+
+  public void alterBeforeNextPageTestPattern() {
```
Renamed and pushed

dkropachev commented on scylla-cdc-lib/src/test/java/com/scylladb/cdc/lib/AlterDropColIT.java:
```diff
@@ -0,0 +1,105 @@
+  }
+
+  @Test
+  public void alterBeforeNextPageTestPattern() {
```
is it done ? can i review it ?

Bouncheck replied:
I thought this was just a question, I'll add it then

Bouncheck replied:
It's done
